core/src/main/java/com/oath/oak/ValueUtilsImpl.java

```diff
@@ -62,44 +62,53 @@ public void unsafeBufferToIntArrayCopy(ByteBuffer bb, int srcPosition, int[] dst
     }
 
     @Override
-    public <V> ValueResult put(Chunk<?, V> chunk, Chunk.LookUp lookUp, V newVal, OakSerializer<V> serializer,
-                               MemoryManager memoryManager) {
+    public <V> ValueResult put(Chunk<?, V> chunk, EntrySet.LookUp lookUp, V newVal, OakSerializer<V> serializer,
+        MemoryManager memoryManager, InternalOakMap internalOakMap) {
         ValueResult result = lockWrite(lookUp.valueSlice, lookUp.version);
         if (result != TRUE) {
             return result;
         }
-        Slice s = innerPut(chunk, lookUp, newVal, serializer, memoryManager);
+        Slice s = lookUp.valueSlice;
+        int capacity = serializer.calculateSize(newVal);
+        if (capacity + getHeaderSize() > s.getByteBuffer().remaining()) {
+            s = moveValue(chunk, lookUp, memoryManager, internalOakMap, newVal);
+            if (s == null) {
+                // rebalance was needed or the entry was updated by someone else, need to retry
+                unlockWrite(lookUp.valueSlice);
+                return RETRY;
+            }
+        } else {
+            s = innerPut(chunk, lookUp, newVal, serializer, memoryManager);
+        }
         unlockWrite(s);
         return TRUE;
     }
 
-    private <V> Slice innerPut(Chunk<?, V> chunk, Chunk.LookUp lookUp, V newVal, OakSerializer<V> serializer,
+    private <V> Slice innerPut(Chunk<?, V> chunk, EntrySet.LookUp lookUp, V newVal, OakSerializer<V> serializer,
                                MemoryManager memoryManager) {
         Slice s = lookUp.valueSlice;
         int capacity = serializer.calculateSize(newVal);
         if (capacity + getHeaderSize() > s.getByteBuffer().remaining()) {
-            s = moveValue(chunk, lookUp, capacity, memoryManager);
+            throw new IllegalStateException(); // sanity check
```

Comment:
If I understand your first comment correctly you suggest just to call moveValue instead of throwing exception. This can not be resolved that simple, as there is a lock that needs to be released and returned Slice can be NULL. Regarding your second comment I changed that capacity is got as input parameter.

Reply:
I suggested three changes that solved this issue and prevent code repetition.

A. change the unlock to use the original slice inside lookup.
This way the correct slice is unlocked all the time.
In addition, if the slice is null, return the appropriate value.
here: https://github.com/yahoo/Oak/pull/84/files/6d6c60bff4ecf5daa525ae234e180e329a9ca5e1#r393563042
```java
unlockWrite(lookUp.valueSlice);
// rebalance was needed or the entry was updated by someone else, need to retry
return s != null ? TRUE : RETRY
```

B. Change back the call to `moveValue()` (this comment). 
```java
return moveValue(chunk, lookUp, capacity, memoryManager, newVal);
```

C. Remove the if/else here: https://github.com/yahoo/Oak/pull/84/files/6d6c60bff4ecf5daa525ae234e180e329a9ca5e1#r393553740

D. This exception is redundant anyway as it is a sanity check and should never happen.